format/format.go

```diff
@@ -471,10 +471,19 @@ func (w *writer) insertComments(comments []*ast.Comment, loc *ast.Location) []*a
 
 func (w *writer) writeBody(body ast.Body, comments []*ast.Comment) []*ast.Comment {
 	comments = w.insertComments(comments, body.Loc())
-	offset := 0
 	for i, expr := range body {
-		if i > 0 && expr.Location.Row-body[i-1].Location.Row-offset > 1 {
-			w.blankLine()
+		// Insert a blank line in before the expression if it was not right
+		// after the previous expression.
```

Comment:
I don't understand this comment 😅 AFAICS in the tests below, blank lines are only removed? Is there a test case that can help me understand the circumstances for when one is inserted?

Reply:
Consider the following code (annotated with line numbers):

```
1|  every i in ls {
2|      ...
3|  }
4|  input.enabled
```

If we're formatting `input.enabled`, we need to decide if we want to insert a blank line before it.  In this case, `lastRow` is 3 (last row of the previous expression).  `input.enabled` follow "right after" the previous expression (on line 4).

If instead we get:

```
1|  every i in ls {
2|      ...
3|  }
4|
5|  input.enabled
```

`lastRow` will again be 3 but `input.enabled` does not follow "right after" the previous expression, so we do want to insert in a blank line, to preserve the logical sections that the user had in the code (sort of like `gofmt` does).

---

The previous code was _always_ inserting a blank line, that why the diff looks like that.

A good example is `test.rego` in the tests:

- [input file](https://github.com/jaspervdj/opa/blob/fix/format-blank-line/format/testfiles/test.rego#L140)
- [new output -- sections are preserved](https://github.com/jaspervdj/opa/blob/fix/format-blank-line/format/testfiles/test.rego#L140)
- [old output -- blank lines are inserted after e.g. `j`](https://github.com/open-policy-agent/opa/blob/main/format/testfiles/test.rego.formatted#L160)

Reply:
Ah! That makes perfect sense. Thanks for the detailed explanation 👍